### DISCUSSION TITLE
Changing SPI configuration during execution

### DISCUSSION THREAD

**Discussion by baptiste-cartier** [zephyrproject-rtos/zephyr#49955] [Q&A]
Hello,

On one of our custom boards, based on the STM32L4, we have a SPI-NOR Flash using Full Duplex (4 wires) SPI communication and a LSM303AGR accelerometer using Half Duplex (3 wires) SPI communication. They both use the same SPI instance and the pins used are the same. We would like to port our application with Zephyr, but as far as I understood, Zephyr does not support dynamic reconfiguration of peripheral through the DTS. Did I miss something ?

What work should be done to support dynamic reconfiguration of this SPI configuration? I guess we could bypass entirely the Device Tree and access the SPI driver but that seems against the philosophy of Zephyr.

We also noticed that the "DTS-ST HAL interface" (in spi_ll_stm32.c) does not support Half Duplex (it does not check the duplex bit in the operation field) but a fix should not be difficult to implement.

Best regards,
BCartier

**Comment by yashi** [comment]
Do you need to re-configure the pins? Assuming you have the 3-wire fix in the driver, it should just work if the chip select pins (or slave select) are connected to them, no?  The devices are using SPI after all.

If you need dynamic reconfigure, what you need is this
- #40385

The [doc](https://docs.zephyrproject.org/latest/hardware/pinctrl/index.html#dynamic-pin-control) says:

> Dynamic pin control should only be used on devices that have not been initialized. Changing pin configurations while a device is operating may lead to unexpected behavior. Since Zephyr does not support device de-initialization yet, this functionality should only be used during early boot stages.
